pkg/metrics/providers/keptn.go

```diff
@@ -0,0 +1,252 @@
+package providers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+)
+
+// api version for the Keptn Metric CRDs
+const (
+	apiVersion = "v1beta1"
+
+	keptnMetricsResourceName = "keptnmetrics"
+	analysisResourceName     = "analyses"
+)
+
+var keptnMetricsResource = schema.GroupVersionResource{
+	Group:    "metrics.keptn.sh",
+	Version:  apiVersion,
+	Resource: keptnMetricsResourceName,
+}
+
+var analysisResource = schema.GroupVersionResource{
+	Group:    "metrics.keptn.sh",
+	Version:  apiVersion,
+	Resource: analysisResourceName,
+}
+
+type queryObject struct {
+	GroupVersionResource schema.GroupVersionResource
+	ResourceName         string
+	DurationString       string
+	Namespace            string
+	Arguments            map[string]interface{}
+}
+
+type KeptnProvider struct {
+	client          dynamic.Interface
+	analysisTimeout time.Duration
+}
+
+func NewKeptnProvider(cfg *rest.Config) (*KeptnProvider, error) {
+	if cfg == nil {
+		return nil, errors.New("could not initialize KeptnProvider: no KubeConfig provided")
+	}
+	client, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize KeptnProvider: %w", err)
+	}
+	return &KeptnProvider{
+		client:          client,
+		analysisTimeout: 10 * time.Second,
```

Comment:
has this been chosen randomly? how does this play with the Canary analysis's interval?

Reply:
This is the maximum time the provider will wait for the result of a Keptn Analysis to be available - I added this to prevent the provider from blocking the flagger controller indefinitely in case the created Analysis cannot be completed. 
I chose this value as this should be sufficient in most cases for the created Keptn Analysis to be reconciled by the Keptn Analysis controller, however I can also adjust this If you'd like, or if there are some guidelines on how long a provider might take to retrieve a value.